### PR TITLE
Prefix global javascript functions with sub-namespaces

### LINKED ISSUE #15299: Building OpenCV.js with projectPoints + solvePnP
Porting to JavaScript: "Cannot register public name 'projectPoints' twice"

I did the following:

1. git clone https://github.com/opencv/opencv.git

2. git clone https://github.com/opencv/opencv_contrib.git

3. I added the following in def get_build_flags(self) of opencv/platforms/js/build_js.py:
        flags += "-s USE_PTHREADS=0 "

4. I enabled the build flag in def get_cmake_cmd(self): of opencv/platforms/js/build_js.py:`-DBUILD_opencv_calib3d` set to `ON`

5. I added the following def get_cmake_cmd(self): of opencv/platforms/js/build_js.py:`-DOPENCV_EXTRA_MODULES_PATH=~/opencv_contrib/modules`

6. I appended `js` inside ocv_define_module at the end of WRAP list of opencv/modules/calib3d/CMakeLists.txt. In opencv/modules/features2d/CMakeLists.txt and opencv_contrib/modules/aruco/CMakeLists.txt also I added  "js" parameter in ocv_define_module.

7. I added `solvePnP`and 'projectPoints' in the calib3d module in the opencv/modules/js/src/embindgen.py 

calib3d = {'': ['findHomography','calibrateCameraExtended', 'drawFrameAxes', 
'getDefaultNewCameraMatrix', 'initUndistortRectifyMap', 'solvePnP','projectPoints']}

8. I added the calib3d module to the makeWhiteList in the opencv/modules/js/src/embindgen.py

white_list = makeWhiteList([core, imgproc, objdetect, video, dnn, features2d, photo, aruco, calib3d])

9. I added "using namespace aruco;" in the opencv/modules/js/src/core_bindings.cpp

10. I built OpenCV.js using the following command: 
sudo python ./platforms/js/build_js.py build_js --emscripten_dir=${EMSCRIPTEN} --clean_build_dir --build_test

Before adding these wrappers, it compiled perfectly without errors. Now in my tests.html I have the following message:
Downloading...
tests.html:61 Running...
tests.html:61 Exception thrown, see JavaScript console
opencv.js:24 Uncaught 
BindingError
message: "Cannot register public name 'projectPoints' twice"
name: "BindingError"
stack: "BindingError: Cannot register public name 'projectPoints' twice↵    at BindingError.<anonymous> (http://localhost/opencv_js/docs/bin/opencv.js:24:9692639)↵    at new BindingError (eval at createNamedFunction (http://localhost/opencv_js/docs/bin/opencv.js:24:9692348), <anonymous>:4:34)↵    at throwBindingError (http://localhost/opencv_js/docs/bin/opencv.js:24:9697766)↵    at exposePublicSymbol (http://localhost/opencv_js/docs/bin/opencv.js:24:9703298)↵    at __embind_register_function (http://localhost/opencv_js/docs/bin/opencv.js:24:9727605)↵    at wasm-function[802]:8129↵    at wasm-function[75]:373↵    at Module.globalCtors (http://localhost/opencv_js/docs/bin/opencv.js:24:9762125)↵    at func (http://localhost/opencv_js/docs/bin/opencv.js:24:9597365)↵    at callRuntimeCallbacks (http://localhost/opencv_js/docs/bin/opencv.js:24:13606)"
__proto__: Error

So seems like the overload functions are preventing me from porting them to JavaScript.

Any suggestions please of how I can fix it?

Thanks in advance for your help.

### PR DESCRIPTION
**Pull request by keroiber**
### Description
Fixes #15299
Both the global namespace and the `fisheye` namespace contain the function `projectPoints`. The bindings generator exports both functions with the original name globally. This solution adds the namespace as prefix to global function names, resulting in distict function names. In this case: `projectPoints` and `fisheye_projectPoints`.

### Incompatibilities
All functions in namespaces will get a prefix in js. From the modules that are exported by default the dnn module is affected.

Personally I would prefer to make javscript objects for namespaces and add _all_ enums, classes and functions there - but that would be a bigger change.

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
build_image:Docs=docs-js:18.04
buildworker:Docs=linux-1,linux-2
build_image:Custom=javascript
buildworker:Custom=linux-4,linux-6
```

### CODE REVIEW

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

This change also replaces names of "dnn" module calls which can break existed user code. So it can't be accepted in the current form.

Could you please reimplement `namespaces_dict` from Java binding?
Prefix is applied to fisheye through [config file](https://github.com/opencv/opencv/blame/4.5.3/modules/calib3d/misc/java/gen_dict.json#L5-L7)

**Comment by keroiber** [comment]
@alalek as I mentioned in the PR I am aware of that problem and didn't expect the PR to be accepted. Thanks for the `namespaces_dict` hint, that was exactly what I was searching for. I will implement it that way.

Nevertheless I still would prefer to see the namespaces reflected in the js interface as described in the PR. That can avoid future name clashes. Assume a new function in dnn that has a namesake in another module. This cannot be solved with the namespace dictionary approach.
Would there be a chance to get that? Maybe as an alternative or additional js build result?

**Comment by alalek** [comment]
I take a look on original issue one more time.
JS functions are filtered through whitelist (to avoid very large build artifacts).
Example of such configuration is here: https://github.com/opencv/opencv/blob/master/platforms/js/opencv_js.config.py
Perhaps we should improve that to properly handle namespaces too.

P.S. need to check if embindgen could emulate "namespaces" (to avoid "ns_" prefixes).

**Comment by keroiber** [comment]
> Example of such configuration is here: https://github.com/opencv/opencv/blob/master/platforms/js/opencv_js.config.py

That's where I intend to add the `namespaces_dict` - this is the correct place, isn't it?

> P.S. need to check if embindgen could emulate "namespaces" (to avoid "ns_" prefixes).

I also didn't check that til now. It seens that embind doesn't support that, see also [this issue](https://github.com/emscripten-core/emscripten/issues/5874)

**Comment by alalek** [comment]
Yes. this is correct place.
Currently binding generator processes 2 instances from `cv` and `cv::fisheye` namespaces.

Probably we can fix filtering process (and exclude `cv::fisheye`) to resolve the original issue. It would be a quick fix, so we can return to nested namespaces problem later (looks like that patch would be large enough).

**Comment by alalek** [comment]
Implemented filtering based on `namespace_prefix_override`.
Rebased to 3.4 branch.

File: platforms/js/opencv_js.config.py [hunk 1]
```diff
@@ -55,8 +55,26 @@
...
+
+        # cv::fisheye namespace
+        'fisheye_initUndistortRectifyMap',
+        'fisheye_projectPoints',
+    ],
+}
 
 white_list = makeWhiteList([core, imgproc, objdetect, video, dnn, features2d, calib3d])
+
+# namespace_prefix_override['dnn'] = ''  # compatibility stuff (enabled by default)
```
Comment (keroiber) [inline platforms/js/opencv_js.config.py:80]: Wouldn't it be clearer to define the compatibility behaviour here and leave namespace_prefix_override empty in embindgen.py?

File: platforms/js/opencv_js.config.py [hunk 2]
```diff
@@ -55,8 +55,26 @@
...
+        'solvePnP',
+        'solvePnPRansac',
+        'solvePnPRefineLM',
+        'projectPoints',
+
+        # cv::fisheye namespace
+        'fisheye_initUndistortRectifyMap',
+        'fisheye_projectPoints',
+    ],
+}
```
Comment (keroiber) [inline platforms/js/opencv_js.config.py:76]: Now this is the only list in the file that's formatted this way

**Review by keroiber — COMMENTED** [review]
Looks good

File: platforms/js/opencv_js.config.py (hunk 2 above)
Comment (alalek) [reply platforms/js/opencv_js.config.py:76]: This file is a sample of opencv.js configuration.
Users of custom opencv.js builds already have own configurations, so we need guarantee compatibility for them.

File: platforms/js/opencv_js.config.py (hunk 1 above)
Comment (alalek) [reply platforms/js/opencv_js.config.py:80]: This file is a sample of opencv.js configuration.
Users of custom opencv.js builds already have own configurations, so we need guarantee compatibility for them.

File: platforms/js/opencv_js.config.py (hunk 2 above)
Comment (alalek) [reply platforms/js/opencv_js.config.py:76]: 1 item per-line formatting is for easy merging between branches (it is already partially used on the master/4.x branch).

**Review by alalek — APPROVED** [review]

**Comment by keroiber** [comment]
@alalek thank you for finishing the compatibility parts

**Pull request merged**